lib/srv/app/server_test.go

```diff
@@ -411,6 +425,97 @@ func TestAWSConsoleRedirect(t *testing.T) {
+	testhttp.Config.TLSConfig = &tls.Config{Time: clockwork.NewFakeClock().Now}
+	testhttp.Start()
+
+	app, err := types.NewAppV3(types.Metadata{
+		Name:   "foo",
+		Labels: staticLabels,
+	}, types.AppSpecV3{
+		URI:           testhttp.URL,
+		PublicAddr:    "foo.example.com",
+		DynamicLabels: types.LabelsToV2(dynamicLabels),
+	})
+	require.NoError(t, err)
+
+	requestEventsReceived := atomic.NewUint64(0)
+	serverStreamer, err := events.NewCallbackStreamer(events.CallbackStreamerConfig{
+		Inner: events.NewDiscardEmitter(),
+		OnEmitAuditEvent: func(_ context.Context, _ libsession.ID, event apievents.AuditEvent) error {
+			if event.GetType() == events.AppSessionRequestEvent {
+				requestEventsReceived.Inc()
+
+				expectedEvent := &apievents.AppSessionRequest{
+					Metadata: apievents.Metadata{
+						Type: events.AppSessionRequestEvent,
+						Code: events.AppSessionRequestCode,
+					},
+					AppMetadata: apievents.AppMetadata{
+						AppURI:        app.Spec.URI,
+						AppPublicAddr: app.Spec.PublicAddr,
+						AppName:       app.Metadata.Name,
+					},
+					StatusCode: 200,
+					Method:     "GET",
+					Path:       "/",
+				}
+				require.Empty(t, cmp.Diff(
+					expectedEvent,
+					event,
+					cmpopts.IgnoreTypes(apievents.ServerMetadata{}, apievents.SessionMetadata{}, apievents.UserMetadata{}, apievents.ConnectionMetadata{}),
+					cmpopts.IgnoreFields(apievents.Metadata{}, "ID", "ClusterName", "Time"),
+					cmpopts.IgnoreFields(apievents.AppSessionChunk{}, "SessionChunkID"),
+				))
+			}
+
+			return nil
+		},
+	})
+	require.NoError(t, err)
+
+	s := SetUpSuiteWithConfig(t, suiteConfig{
+		ServerStreamer: serverStreamer,
+		Apps:           types.Apps{app},
+	})
+	// make a request to generate events.
+	s.checkHTTPResponse(t, s.clientCertificate, func(_ *http.Response) {
+		searchEvents, _, err := s.authServer.AuditLog.SearchEvents(time.Time{}, time.Now().Add(time.Minute), "", []string{events.AppSessionChunkEvent}, 10, types.EventOrderDescending, "")
+		require.NoError(t, err)
+		require.Len(t, searchEvents, 1)
+
+		expectedEvent := &apievents.AppSessionChunk{
+			Metadata: apievents.Metadata{
+				Type: events.AppSessionChunkEvent,
+				Code: events.AppSessionChunkCode,
+			},
+			AppMetadata: apievents.AppMetadata{
+				AppURI:        app.Spec.URI,
+				AppPublicAddr: app.Spec.PublicAddr,
+				AppName:       app.Metadata.Name,
+			},
+		}
+		require.Empty(t, cmp.Diff(
+			expectedEvent,
+			searchEvents[0],
+			cmpopts.IgnoreTypes(apievents.ServerMetadata{}, apievents.SessionMetadata{}, apievents.UserMetadata{}, apievents.ConnectionMetadata{}),
+			cmpopts.IgnoreFields(apievents.Metadata{}, "ID", "ClusterName"),
+			cmpopts.IgnoreFields(apievents.AppSessionChunk{}, "SessionChunkID"),
+		))
+
+		require.Eventually(t, func() bool {
+			return requestEventsReceived.Load() == 1
+		}, 500*time.Millisecond, 50*time.Millisecond, "app.request event not generated")
+	})
```

Comment:
I've made the change, but the `requestEventsReceived` block (L513) needs to be inside the `checkHTTPResponse`. If not, the event never gets recorded into the variable. I'm not 100% sure why, but it seems that because the `checkHTTPResponse` closes the server, and since this test is using the Sync recorder, the server cancels the emit event.